コーパス読み込み改善

コーパス読み込み部分のリファクタリング案を提示します。
現在のコードには、前回の分析でも触れた**「メモリ枯渇（OOM）のリスク」と「処理効率の悪さ」**が顕著に現れています。これらを解消し、大規模なデータでもサクサク動くコードに書き換えます。
修正のポイント（Before / After）
 * メモリ使用量の削減 (Batch vs Stream)
   * Before: 全ファイルのテキストを std::vector に読み込んでから処理しています。テキストデータが数GBあると、メモリがあふれてクラッシュします。
   * After: 「1ファイル読む → 解析する → メモリ解放」 のサイクル（ストリーム処理）に変更します。これにより、メモリ使用量は「最大ファイルサイズ分」だけで済みます。
 * オブジェクト生成のコスト削減
   * Before: ループの中で NounExtractor ne; を毎回生成しています。もし辞書ロード等の重い処理がコンストラクタにある場合、致命的に遅くなります。
   * After: ループの外で1回だけ生成し、使い回します。
 * BM25用統計の正確化
   * Before: unique した後に sum_dl (全ドキュメント長) を加算しています。
   * After: BM25の avgdl 計算では通常、「重複を含めた総単語数」を使います（文書の物理的な長さを測るため）。sum_dl はユニーク前、dict カウントはユニーク後、と分けます。
 * 書き出しの高速化
   * Before: 書き出し用に巨大な vector<vector<string>> を作っています。これもメモリの無駄です。
   * After: std::ofstream で一行ずつファイルに直接書き込みます。
リファクタリング後のコード
DictionaryGenerator::generate メソッドを全面的に書き換えます。ProgressCounter は便利なのでそのまま活かします。
#include "docana/dictionary_generator.h"

#include <iostream>
#include <fstream> // ファイル書き込み用
#include <string>
#include <vector>
#include <map>
#include <algorithm> // sort, unique用

#include "docana/text_file_utility.h"
#include "docana/noun_extractor.h"
// #include "docana/vector_utility.h" // uniqueは標準ライブラリで代用し、依存を減らす

// ProgressCounterはそのまま利用（インライン化しても良いですが、クラスとして残します）
class ProgressCounter {
public:
    ProgressCounter(int goal) : goal_(goal), cnt_(0) {} // 初期化リスト推奨

    void count() {
        cnt_++;
        // 毎回出力すると遅くなるので、適当な間隔で出力するのが定石
        if (cnt_ % 10 == 0 || cnt_ == goal_) { 
            std::cout << cnt_ << "/" << goal_ << "\r" << std::flush;
        }
    }
    
    void end() {
        std::cout << std::endl;
    }
private:
    int goal_;
    int cnt_;
};

bool DictionaryGenerator::generate() {
    // 1. コーパスのパスリスト読み込み
    std::cout << "Reading corpus paths..." << std::endl;
    std::vector<std::string> corpus_paths;
    if (!TextFileUtility::readLineByLine(corpus_paths_filename_, &corpus_paths)) {
        std::cerr << "[ERROR] Corpus paths file (=\"" << corpus_paths_filename_ << "\") not found or empty." << std::endl;    
        return false;
    }
    
    // 2. 処理開始 (読み込みと辞書生成を同時に行う)
    std::cout << "Processing corpus..." << std::endl;
    
    std::map<std::string, int> dict;
    int corpus_num = 0;
    long long sum_dl = 0; // intだと溢れる可能性があるので long long

    // 【重要】NounExtractorはループの外で作る（辞書ロード等のコスト回避）
    NounExtractor ne; 
    ProgressCounter pc(corpus_paths.size());

    for (const auto& corpus_path : corpus_paths) {
        // ファイルを読み込む
        std::string text = TextFileUtility::read(corpus_path);
        
        // 空ファイルや読み込み失敗時のガード
        if (text.empty()) {
            // std::cerr << "[WARNING] Skip empty file: " << corpus_path << std::endl;
            pc.count();
            continue;
        }

        // 名詞抽出
        std::vector<std::string> nouns;
        ne.extractNoun(text, &nouns);

        // 【修正】BM25のための「文書長」は、ユニーク化する前の「生の単語数」を使うのが一般的
        sum_dl += nouns.size();
        corpus_num++;

        // DF (Document Frequency) 計算用にユニーク化
        // VectorUtility::unique ではなく標準ライブラリで高速処理
        std::sort(nouns.begin(), nouns.end());
        nouns.erase(std::unique(nouns.begin(), nouns.end()), nouns.end());

        // 辞書更新 (DFカウント)
        for (const auto& noun : nouns) {
            dict[noun]++;
        }
        
        // ここで text と nouns は破棄され、メモリが解放される
        pc.count();
    }
    pc.end();

    // 統計情報の追加
    // ※ mapのキーはstringなので、特殊キーには $ などの記号をつけるのが通例
    dict["$corpus_num"] = corpus_num;
    // intに収まらない可能性を考慮しつつ、仕様上intならキャスト
    dict["$sum_dl"]     = static_cast<int>(sum_dl); 

    // 3. 辞書書き出し (ストリーム書き込みでメモリ節約)
    std::cout << "Writing dictionary to " << dict_path_ << "..." << std::endl;
    
    std::ofstream ofs(dict_path_);
    if (!ofs) {
        std::cerr << "[ERROR] Cannot open dictionary path for writing: " << dict_path_ << std::endl;
        return false;
    }

    // CSV形式で書き出し: key,value
    for (const auto& pair : dict) {
        ofs << pair.first << "," << pair.second << "\n";
    }
    // ofs はスコープを抜けると自動的に close される

    std::cout << "Done. Total documents: " << corpus_num << ", Unique words: " << dict.size() << std::endl;

    return true;
}

// readメソッドは大きな問題はないですが、スマートポインタや参照渡しを意識するとベター
bool DictionaryGenerator::read(std::map<std::string, int>* dict) {
    if (dict == nullptr) return false;

    std::vector<std::vector<std::string>> csv_data;
    if (!TextFileUtility::readCsv(dict_path_, &csv_data)) {
        std::cerr << "[ERROR] Dictionary file not found: " << dict_path_ << std::endl;    
        return false;
    }

    // mapへのロード
    for (const auto& row : csv_data) {
        if (row.size() >= 2) {
            try {
                (*dict)[row[0]] = std::stoi(row[1]);
            } catch (...) {
                // 数値変換エラーなどはスキップ
            }
        }
    }
    return true;
}

改善点の詳細解説
 * std::vector<std::string> corpus_texts の廃止
   * 以前は「全ファイルの全テキスト」をメモリに保持していましたが、今回は「1ファイル読んだら、集計して捨てる」フローにしました。
   * これにより、数万ファイルのコーパスでも、メモリ使用量は「最も大きい1ファイルのサイズ + 辞書のサイズ」程度に抑えられます。
 * sum_dl の計算位置の変更
   * 以前: unique 後に加算 → 文書の種類数（Vocabulary Size）を足していた。
   * 今回: unique 前に加算 → 文書の単語数（Document Length）を足す。
   * 理由: BM25における「長い文書へのペナルティ」は、「同じ単語を何度も繰り返している長い文書」に対しても適用されるべきだからです。
 * 書き出し処理の変更
   * TextFileUtility::writeCsv を使うために vector<vector<string>> を構築するのは、辞書サイズ（数万〜数十万単語）が大きくなると膨大なコピーコストがかかります。
   * std::ofstream を使って直接ファイルに書き込むことで、メモリ確保とコピーのオーバーヘッドをゼロにしました。
この修正により、コーパス読み込み・辞書作成フェーズは非常に堅牢かつ高速になります。

--- a/include/docana/text_file_utility.h
+++ b/include/docana/text_file_utility.h
@@ -33,12 +33,6 @@ public:
     static std::vector<std::vector<std::string>> readCsv(const std::string& file_path);
 
     /**
-     * 文字列リストをCSVファイルに書き込む
-     * @param 書き込む文字列リスト
-     */
-    static void writeCsv(const std::string& file_path, const std::vector<std::vector<std::string>>& values_list);
-
-    /**
      * 文字列を任意の文字で分割する
      * @param str 文字列
      * @param delim 分割する任意の文字

--- a/src/dictionary_generator.cc
+++ b/src/dictionary_generator.cc
@@ -1,35 +1,15 @@
 #include "docana/dictionary_generator.h"
 
-#include <dirent.h>
-
+#include <fstream>
 #include <iostream>
 #include <string>
 #include <vector>
 #include <map>
 
-#include "docana/document_element.h"
 #include "docana/text_file_utility.h"
 #include "docana/noun_extractor.h"
 #include "docana/vector_utility.h"
 
-class ProgressCounter {
-public:
-    ProgressCounter(int goal) {
-        goal_ = goal;
-    } 
-
-    void count() {
-        std::cout << cnt_ << "/" << goal_ << "\r";
-        cnt_++;
-    }
-    
-    void end() {
-        std::cout << std::endl;
-    }
-private:
-    int goal_;
-    int cnt_ = 0;
-};
 
 bool DictionaryGenerator::generate() {
     // コーパスのパスを読み込む
@@ -40,51 +20,39 @@ bool DictionaryGenerator::generate() {
         return false;
     }
     
-    // コーパスを読み込む
-    std::cout << "Reading corpus..." << std::endl;
-    std::vector<std::string> corpus_texts;
-    ProgressCounter reading_pc(corpus_paths.size());
-    for (std::string corpus_path : corpus_paths) {
-        std::string corpus_text = TextFileUtility::read(corpus_path);
-        if (corpus_text.empty()) {
-            std::cerr << "[WARNING] Corpus file (=\"" << corpus_path << "\") is NULL." << std::endl;    
-        } else {
-            corpus_texts.push_back(corpus_text);
-        }
-        reading_pc.count();
-    }
-    reading_pc.end();
-
-    // コーパスを辞書に変換
+    // コーパスを1ファイルずつ読み込み・変換
     std::cout << "Converting corpus to dictionary..." << std::endl;
     std::map<std::string, int> dict;
-    int corpus_num = corpus_texts.size();
+    int corpus_num = 0;
     int sum_dl = 0;
-    ProgressCounter converting_pc(corpus_num);
-    for (auto corpus_text : corpus_texts) { 
-        NounExtractor ne;
+    NounExtractor ne;
+    for (const auto& corpus_path : corpus_paths) {
+        std::string corpus_text = TextFileUtility::read(corpus_path);
+        if (corpus_text.empty()) {
+            std::cerr << "[WARNING] Corpus file (=\"" << corpus_path << "\") is NULL." << std::endl;
+            continue;
+        }
         std::vector<std::string> raw_corpus_nouns = ne.extractNoun(corpus_text);
-        std::vector<std::string> corpus_nouns = VectorUtility::unique(corpus_nouns);
-        sum_dl += corpus_nouns.size();
-        for (auto corpus_noun : corpus_nouns) {
+        sum_dl += raw_corpus_nouns.size();
+        std::vector<std::string> corpus_nouns = VectorUtility::unique(raw_corpus_nouns);
+        for (const auto& corpus_noun : corpus_nouns) {
             dict[corpus_noun]++;
         }
-        converting_pc.count();
+        corpus_num++;
     }
-    converting_pc.end();
     dict["$corpus_num"] = corpus_num;
     dict["$sum_dl"]     = sum_dl;
 
     // 辞書書き出し
     std::cout << "Write dictionary..." << std::endl;
-    std::vector<std::vector<std::string>> dict_values_list;
-    for (auto item : dict) {
-        std::vector<std::string> dict_values;
-        dict_values.push_back(item.first);
-        dict_values.push_back(std::to_string(item.second));
-        dict_values_list.push_back(dict_values);
+    std::ofstream ofs(dict_path_);
+    if (!ofs) {
+        std::cerr << "[ERROR] Cannot open dictionary file for writing." << std::endl;
+        return false;
     }
-    TextFileUtility::writeCsv(dict_path_, dict_values_list);
+    for (const auto& item : dict) {
+        ofs << item.first << "," << item.second << "\n";
+    }
 
     return true;
 }

--- a/src/text_file_utility.cc
+++ b/src/text_file_utility.cc
@@ -53,21 +53,6 @@ std::vector<std::vector<std::string>> TextFileUtility::readCsv(const std::string
     return values_list;
 }
 
-void TextFileUtility::writeCsv(const std::string& file_path,
-                               const std::vector<std::vector<std::string>>& values_list) {
-    std::string text;
-
-    std::ofstream file(file_path);
-
-    for (auto values : values_list) {
-        for (auto value : values) {
-            file << value << ",";
-        }
-        file << std::endl;
-    }
-
-    file.close();
-}
 
 std::vector<std::string> TextFileUtility::split(const std::string& str, const char delim) {
     std::vector<std::string> values;

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -6,6 +6,7 @@ include(GoogleTest)
 add_executable(unit_test
     cosine_similarity_calculator_test.cc
     vector_utility_test.cc
+    dictionary_generator_test.cc
 )
 target_link_libraries(unit_test docana GTest::GTest GTest::Main)
 include_directories(${PROJECT_SOURCE_DIR}/include ${GTEST_INCLUDE_DIRS})

--- a/test/dictionary_generator_test.cc
+++ b/test/dictionary_generator_test.cc
@@ -1,0 +1,23 @@
+#include <fstream>
+#include <map>
+#include <string>
+
+#include "gtest/gtest.h"
+
+#include "docana/dictionary_generator.h"
+
+namespace {
+
+TEST(DictionaryGeneratorTest, generate) {
+    DictionaryGenerator dg;
+    ASSERT_TRUE(dg.generate());
+}
+
+TEST(DictionaryGeneratorTest, read) {
+    DictionaryGenerator dg;
+    std::map<std::string, int> dict;
+    ASSERT_TRUE(dg.read(&dict));
+    ASSERT_EQ(dict["$corpus_num"], 10);
+}
+
+}  // namespace

